core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,236 @@
+          val p = unCompletedPromises(ret.context)
+          // get the context
+          promiseContext = p.ctx
+          ret.returnValue match {
+            case throwable: Throwable =>
+              // if an error is returned
+              p.setException(throwable)
+            case _ =>
+              // otherwise, we put the return value
+              // and invoke the callback.
+              p.setValue(ret.returnValue.asInstanceOf[p.returnType])
+          }
+          // the promise is resolved
+          unCompletedPromises.remove(ret.context)
+        }
+
+        // check if the return value belongs to group promises
+        for (i <- unCompletedGroupPromises) {
+          if (i.takeReturnValue(ret)) {
+            // all return values are collected, set the context
+            promiseContext = i.promise.ctx
+            // invoke callback
+            i.invoke()
+            // the group promise is resolved
+            unCompletedGroupPromises.remove(i)
+          }
+        }
+
+      // handle root synchronous promise
+      case PromiseInvocation(
+            ctx: RootPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null) {
+          // if there is no other executing synchronous promise,
+          // execute this one
+          registerSyncPromise(ctx, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // handle synchronous promise created by other promise
+      case PromiseInvocation(
+            ctx: ChildPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null || ctx.root == syncPromiseRoot) {
+          // if there is no other executing synchronous promise,
+          // or this promise has same root context (chain of promises),
+          // execute this one.
+          registerSyncPromise(ctx.root, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // trivial case
+      case p: PromiseInvocation =>
+        // set context
+        promiseContext = p.context
+        try {
+          // execute it
+          promiseHandler(p.call)
+        } catch {
+          case e: Throwable =>
+            // if error occurs, return it to the sender.
+            returning(e)
+        }
+    }
+
+    // execute queued sync promises
+    tryInvokeNextSyncPromise()
+  }
+
+  // send a control message to another actor, and keep the handle.
+  def schedule[T](cmd: PromiseBody[T], on: ActorVirtualIdentity): Promise[T] = {
+    val ctx = mkPromiseContext()
+    promiseID += 1
+    controlOutputPort.sendTo(on, PromiseInvocation(ctx, cmd))
+    val promise = WorkflowPromise[T](promiseContext)
+    unCompletedPromises(ctx) = promise
+    promise
```

Comment:
move line 139-141 before sending the message out. Then the first 2 and these 3 lines are actually `createLocalPromise`. We need a better name for `createLocalPromise`. How about just createPromise?

Reply:
Since we need both `ctx` and `promiseContext` in this function, I don't want to change it.